core/services/ocr2/plugins/ocr2keeper/evm21/feed_lookup.go

```diff
@@ -118,90 +118,94 @@ func (r *EvmRegistry) feedLookup(ctx context.Context, upkeepResults []ocr2keeper
 	}
 
-	mercuryBytes, err := r.checkCallback(ctx, values, lookup)
+	state, retryable, mercuryBytes, err := r.checkCallback(ctx, values, lookup)
 	if err != nil {
 		r.lggr.Errorf("[FeedLookup] upkeep %s block %d checkCallback err: %v", lookup.upkeepId, lookup.block, err)
+		checkResults[i].Retryable = retryable
+		checkResults[i].PipelineExecutionState = uint8(state)
 		return
 	}
 	r.lggr.Infof("[FeedLookup] checkCallback mercuryBytes=%s", hexutil.Encode(mercuryBytes))
 
-	needed, performData, failureReason, _, err := r.packer.UnpackCheckCallbackResult(mercuryBytes)
+	state, needed, performData, failureReason, _, err := r.packer.UnpackCheckCallbackResult(mercuryBytes)
 	if err != nil {
 		r.lggr.Errorf("[FeedLookup] upkeep %s block %d UnpackCheckCallbackResult err: %v", lookup.upkeepId, lookup.block, err)
+		checkResults[i].PipelineExecutionState = uint8(state)
 		return
 	}
 
 	if failureReason == uint8(UpkeepFailureReasonMercuryCallbackReverted) {
-		upkeepResults[i].IneligibilityReason = uint8(UpkeepFailureReasonMercuryCallbackReverted)
+		checkResults[i].IneligibilityReason = uint8(UpkeepFailureReasonMercuryCallbackReverted)
 		r.lggr.Debugf("[FeedLookup] upkeep %s block %d mercury callback reverts", lookup.upkeepId, lookup.block)
 		return
 	}
 
 	if !needed {
-		upkeepResults[i].IneligibilityReason = uint8(UpkeepFailureReasonUpkeepNotNeeded)
+		checkResults[i].IneligibilityReason = uint8(UpkeepFailureReasonUpkeepNotNeeded)
 		r.lggr.Debugf("[FeedLookup] upkeep %s block %d callback reports upkeep not needed", lookup.upkeepId, lookup.block)
 		return
 	}
 
-	upkeepResults[i].IneligibilityReason = uint8(UpkeepFailureReasonNone)
-	upkeepResults[i].Eligible = true
-	upkeepResults[i].PerformData = performData
+	checkResults[i].IneligibilityReason = uint8(UpkeepFailureReasonNone)
+	checkResults[i].Eligible = true
+	checkResults[i].PerformData = performData
 	r.lggr.Infof("[FeedLookup] upkeep %s block %d successful with perform data: %s", lookup.upkeepId, lookup.block, hexutil.Encode(performData))
 }
 
 // allowedToUseMercury retrieves upkeep's administrative offchain config and decode a mercuryEnabled bool to indicate if
 // this upkeep is allowed to use Mercury service.
-func (r *EvmRegistry) allowedToUseMercury(opts *bind.CallOpts, upkeepId *big.Int) (bool, error) {
+func (r *EvmRegistry) allowedToUseMercury(opts *bind.CallOpts, upkeepId *big.Int) (state PipelineExecutionState, retryable bool, allow bool, err error) {
 	allowed, ok := r.mercury.allowListCache.Get(upkeepId.String())
 	if ok {
-		return allowed.(bool), nil
+		return NoPipelineError, false, allowed.(bool), nil
 	}
 
 	cfg, err := r.registry.GetUpkeepPrivilegeConfig(opts, upkeepId)
 	if err != nil {
-		return false, fmt.Errorf("failed to get upkeep privilege config for upkeep ID %s: %v", upkeepId, err)
+		return RpcFlakyFailure, true, false, fmt.Errorf("failed to get upkeep privilege config for upkeep ID %s: %v", upkeepId, err)
```

Comment:
this can be a RPC failure or the contract reverts. not sure if it's worthy to parse the error string to differentiate.

Reply:
add a comment and let's handle in a later PR